src/java/voldemort/server/VoldemortServer.java

```diff
@@ -520,21 +524,60 @@ public void goOffline() {
     }
 
     public void goOnline() {
-        getMetadataStore().setOfflineState(false);
-        List<StorageEngine<ByteArray, byte[], byte[]>> storageEngines =
-                storageService.getStoreRepository().getStorageEnginesByClass(ReadOnlyStorageEngine.class);
-
-        if (storageEngines.size() > 0) {
-            logger.info("Will attempt to removeRemoteObsoleteState() for " + storageEngines.size() + " Read-Only stores.");
-            for (StorageEngine storageEngine: storageEngines) {
-                StoreVersionManager storeVersionManager =
-                        (StoreVersionManager) storageEngine.getCapability(StoreCapabilityType.DISABLE_STORE_VERSION);
-                try {
-                    storeVersionManager.removeRemoteObsoleteState();
-                } catch (Exception e) {
-                    logger.error("Failed to removeRemoteObsoleteState() for store: " + storageEngine.getName(), e);
-                }
-            }
+        if (getVoldemortConfig().getHighAvailabilityStateAutoCleanUp()) {
+            final int asyncRequestId = asyncService.getUniqueRequestId();
+            logger.info("Since " + VoldemortConfig.PUSH_HA_STATE_AUTO_CLEANUP + "=true, async task ID " +
+                        asyncRequestId + " will be launched to ensure the HA state is cleaned up.");
+            asyncService.submitOperation(
+                    asyncRequestId,
+                    new AsyncOperation(asyncRequestId, "HA state clean up.") {
```

Comment:
Sure, I can. It still seems contentious whether this should be asynchronous in the first place or not, however. Let's settle this before making any further changes.